docs/avansat/suffix-array-tree.md

```diff
@@ -0,0 +1,136 @@
+
+
+| Indicele de la care începe rotația |     Rotația     |
+|:----------------------------------:|----------------:|
+|               $6$                  |    $\text{\$banana}$  |
+|               $5$                  |    $\text{a\$banan}$  |
+|               $3$                  |    $\text{ana\$ban}$  |
+|               $1$                  |    $\text{anana\$b}$  |
+|               $0$                  |    $\text{banana\$}$  |
+|               $4$                  |    $\text{na\$bana}$  |
+|               $2$                  |    $\text{nana\$ba}$  |
+
+!!!info "Observație" 
+    Se observa că ordinea este identică dupa ce ștergem primul element din tablou. Rotația ce începe cu santinela va fi mereu prima și nu ne modifică rezultatul.
+## Construire
+Sortarea va fi realizată în $\lceil{log_2(N)}\rceil$ pași, unde $N$ este mărimea șirului, la pasul $i$ având sortate rotațiile dupa prefixele lor de lungime $2^i$. Pentru a ne ușura munca introducem conceptul de clase de echivalență, fiecarui șir fiindu-i atribuit un număr natural, fie el $C_i$ pentru rotația ce începe din $i$, pe care le recalculăm la fiecare iterație a sortării. Aceste clase de echivalență au următoarele proprietăți : dacă un șir este mai mic lexicografic decât altul atunci are o clasă de echivalență mai mică iar dacă două șiruri sunt egale atunci au aceeași clasă de echivalență. În plus, numarul de clase de echivalență folosite trebuie să fie minim. Bazându-ne pe aceste proprietăți, la pasul $i$ putem sorta prefixele rotațiilor folosind o metodă similară cu radixsort : împărțim prefixul de lungime $2^i$ în două bucăți de lungime $2^{i-1}$ și le sortăm întai după bucata din dreapta apoi după bucata din stânga. Mare atenție : aceste sortări trebuie să nu schimbe ordinea relativă în caz de egalitate, altfel se duce totul de râpă. Astfel, din proprietatea $3$ a claselor de echivalență știim că nu vor depăși niciodată valoarea $N$ și putem folosi metode de sortare în $O(N)$ precum counting sort sau, ce recomand eu, doar ținem un vector de vectori unde $V_i = $ vectorul indicilor sufixelor ale căror clasă de echivalență pe care o comparăm este $i$. Astfel, complexitatea de timp este $O(Nlog_2{N})$ iar cea de spațiu este $O(N)$.
+
+## Câteva detalii de implementare
+
+Pentru a obține clasele jumătăților prefixelor este necesar să lucrăm cu indici modulo $N$. Spre exemplu, dacă $N = 5$, suntem la pasul $2$ și ne dorim să obținem clasele pentru jumătățile corespunzătoare prefixului rotației $3$ atunci acestea vor fi clasa rotației $3$ în pasul $1$ respectiv clasa lui $3 + 2 ^ 1 \equiv 0 \pmod{N}$ în pasul $1$. Este lăsat ca demonstrație pentru cititor de ce algoritmul rămâne corect și după pasul $\lceil{log_2(N)}\rceil$ în care este posibil ca jumătățile să se intersecteze. Vă puteți testa implementarea [aici](https://judge.yosupo.jp/problem/suffixarray) și aveți modelul autorului [aici](https://github.com/MONcalc/Surse_roalgo/blob/main/suffix_array.cpp).
```

Comment:
Poti pune implementarea direct in articol, si de asemenea cred ca se poate ajusta un pic exprimarea "și aveți solutia model a autorului"